py/server/deephaven/table_listener.py

```diff
@@ -442,3 +445,173 @@ def listen(t: Table, listener: Union[Callable, TableListener], description: str
+        (1) a callable (e.g. function) or
+        (2) an instance of MergedTableListener type which provides a "process" method.
+        The callable or the process method must have zero parameters.
+
+        Note: Don't do table operations in the listener. Do them beforehand, and add the results as dependencies.
+
+        Args:
+            listener_recorders (Sequence[ListenerRecorder]): listener recorders that record the table updates for the
+                merged listener to process.
+            listener (Union[Callable, MergedListener]): listener to process table updates from the listener recorders
+            description (str, optional): description for the UpdatePerformanceTracker to append to the listener's entry
+            dependencies (Union[Table, Sequence[Table]]): tables that must be satisfied before the listener's execution.
+                A refreshing table is considered to be satisfied if all possible updates to the table have been processed
+                in the current update graph cycle. A static table is always considered to be satisfied. If a specified
+                table is refreshing, it must belong to the same update graph as the table being listened to. Default is
+                None.
+
+                Dependencies ensure that the listener can safely access the dependent tables during its execution. This
+                mainly includes reading the data from the tables. While performing operations on the dependent tables in
+                the listener is safe, it is not recommended because reading or operating on the result tables of those
+                operations may not be safe. It is best to perform the operations on the dependent tables beforehand,
+                and then add the result tables as dependencies to the listener so that they can be safely read in it.
+
+        Raises:
+            DHError
+        """
+        if len(listener_recorders) < 2:
+            raise DHError(message="MergedListener must have at least two listener recorders.")
+
+        self.dependencies = dependencies
+        self.listener_recorders = listener_recorders
+
+        if isinstance(listener, MergedListener):
+            listener.listener_recorders = listener_recorders
+
+        try:
+            self.merged_listener_adapter = _JPythonMergedListenerAdapter.create(
+                        to_sequence(self.listener_recorders),
+                        to_sequence(self.dependencies),
+                        description,
+                        listener)
+            self.started = False
+        except Exception as e:
+            raise DHError(e, "failed to create a merged listener adapter.") from e
+
+
+    def start(self) -> None:
+        """Start the listener."""
+        if self.started:
+            raise RuntimeError("Attempting to start an already started merged listener..")
+
+        with update_graph.shared_lock(self.listener_recorders[0].table.update_graph):
```

Comment:
I think you need a test here to ensure that you don't get the shared lock if the user already has the exclusive lock. Maybe you want to re-use the auto-locking feature? Same in `stop`.